flink-kubernetes/src/test/java/org/apache/flink/kubernetes/highavailability/KubernetesStateHandleStoreTest.java

```diff
@@ -0,0 +1,382 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.kubernetes.highavailability;
+
+import org.apache.flink.api.common.JobID;
+import org.apache.flink.kubernetes.kubeclient.resources.KubernetesLeaderElector;
+import org.apache.flink.runtime.statehandle.StateHandleStore;
+import org.apache.flink.runtime.statehandle.TestingLongStateHandleHelper.LongRetrievableStateHandle;
+import org.apache.flink.runtime.statehandle.TestingLongStateHandleHelper.LongStateStorage;
+import org.apache.flink.util.function.FunctionUtils;
+
+import org.junit.Before;
+import org.junit.Test;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Comparator;
+import java.util.List;
+import java.util.function.Predicate;
+
+import static org.hamcrest.Matchers.contains;
+import static org.hamcrest.Matchers.containsString;
+import static org.hamcrest.Matchers.is;
+import static org.hamcrest.Matchers.not;
+import static org.junit.Assert.assertThat;
+import static org.junit.Assert.fail;
+
+/**
+ * Tests for {@link KubernetesStateHandleStore} operations.
+ */
+public class KubernetesStateHandleStoreTest extends KubernetesHighAvailabilityTestBase {
+
+	private static final String PREFIX = "test-prefix-";
+	private final String key = PREFIX + JobID.generate();
+	private final Predicate<String> filter = k -> k.startsWith(PREFIX);
+	private final Long state = 12345L;
+
+	private LongStateStorage longStateStorage;
+
+	@Before
+	public void setup() {
+		super.setup();
+		longStateStorage = new LongStateStorage();
+	}
+
+	@Test
+	public void testAdd() throws Exception {
+		new Context() {{
+			runTest(
+				() -> {
+					leaderCallbackGrantLeadership();
+
+					final KubernetesStateHandleStore<Long> store = new KubernetesStateHandleStore<>(
+						flinkKubeClient, LEADER_CONFIGMAP_NAME, longStateStorage, filter, LOCK_IDENTITY);
+					store.add(key, state);
+					assertThat(store.getAll().size(), is(1));
+					assertThat(store.get(key).retrieveState(), is(state));
+				});
+		}};
+	}
+
+	@Test
+	public void testAddAlreadyExistingKey() throws Exception {
+		new Context() {{
+			runTest(
+				() -> {
+					leaderCallbackGrantLeadership();
+
+					final KubernetesStateHandleStore<Long> store = new KubernetesStateHandleStore<>(
+						flinkKubeClient, LEADER_CONFIGMAP_NAME, longStateStorage, filter, LOCK_IDENTITY);
+					store.add(key, state);
+
+					try {
+						store.add(key, state);
+						fail("Exception should be thrown.");
+					} catch (StateHandleStore.AlreadyExistException ex) {
+						final String msg = String.format(
+							"%s already exists in ConfigMap %s", key, LEADER_CONFIGMAP_NAME);
+						assertThat(ex.getMessage(), containsString(msg));
+					}
+					assertThat(longStateStorage.getStateHandles().size(), is(2));
```

Comment:
It is a bit confusing that `longStateStorage` contains two state handles even though the `add` operation failed.

Reply:
Actually, we have two state handles here. One succeed and one failed. Each time we try to store a state handle, a new `LongRetrievableStateHandle` will be created and added to the list.

Maybe I need to rename the method name to `getUsedStateHandles`.